### ISSUE TITLE
FR: add support for multiple connections

### ISSUE THREAD

**Issue by GitMensch** [mridoni/gix#44]
... normally (only?) handled via the `AT` clause, both at connect and when executing, for example:

```cobol
           EXEC SQL AT :DBS
              SELECT PG_BACKEND_PID() INTO :SID
           END-EXEC.
```

Also see #43 

Note: it seems that "upstream" had integrated this after the private fork in ocesql 1.1 [problem initially brought up in #16).
The complete changes for the 1.1 version were committed with [Open-COBOL-ESQL@62f0d15ee92e1a3d1cdc1470e13a84ea68966ce](https://github.com/opensourcecobol/Open-COBOL-ESQL/commit/62f0d15ee92e1a3d1cdc1470e13a84ea68966ce)3 so it seems both reasonable and doable to copy the missing pieces from there (surely there's some adjustment needed).

Note: the missing support for multiple connections is actually a show stopper for my project :-(

**Comment by mridoni** [comment]
There is already some support for this in the main library, what is missing is adding it in the parser code and integrating it (if needed) with the client libraries, I'll surely look into it.

**Comment by GitMensch** [comment]
Thanks! As noted upstream has that implemented in a single commit, it seems reasonable to check this, which potentially allows closing #16, too.

**Comment by mridoni** [comment]
Yes, unfortunately I'll have first to "triage" this, given the differences that currently exist between OceSQL and GixSQL, but even if I can't give a timeline, it can (and will) definitely be done.

**Comment by mridoni** [comment]
Introduced in [v1.0.8](https://github.com/mridoni/gix/releases/tag/v1.0.8)

**Comment by GitMensch** [comment]
Seems buggy, there are errors on unexpected syntax for at least `OPEN`, `FETCH` and `CLOSE` if the `AT :CON` syntax is used.
Do you need code parts for this issue (it seems to be recognizable from gix_esql_parser.yy `opensql:`, `closesql:`, `fetchsql:`)?

**Comment by mridoni** [comment]
Ok, can you provide specific examples in COBOL? Even snippets would be good.

Thanks

**Comment by GitMensch** [comment]
Works (only parsing tested so far) fine:

```cobol
           EXEC SQL AT :DBS
             SELECT FLD1, FLD2 INTO :FLD1, :FLD2 FROM TAB1
             WHERE KEY01 =
                        :KY1A
                     || :KY1B
             LIMIT 1
           END-EXEC.
```

Syntax errors at:

```cobol
           EXEC SQL AT :DBS
              DECLARE CSKEY01N CURSOR FOR
                 SELECT FLD1, FLD2 FROM TAB1
                 WHERE
                    KEY01 >= (
                        :KY1A
                     || :KY1B
                             )
                 ORDER BY KEY01 ASC
           END-EXEC.
      *
           EXEC SQL AT :DBS
              OPEN CSKEY01N
           END-EXEC.
      *
           EXEC SQL AT :DBS
               FETCH CSKEY01N
               INTO
                 :FLD1,
                 :FLD2
           END-EXEC.
```

You may want to reopen this issue.

Note: it looks like the parser stops at the first syntax error (if you remove the `AT :DBS` before the `OPEN` you get to the `FETCH` with the next compile. if it works as I think it should, I highly suggest to try improving error recovery (I'll post a "draft PR" for that next).

**Comment by mridoni** [comment]
This I must check

           EXEC SQL AT :DBS
              DECLARE CSKEY01N CURSOR FOR
                 SELECT FLD1, FLD2 FROM TAB1
                 WHERE
                    KEY01 >= (
                        :KY1A
                     || :KY1B
                             )
                 ORDER BY KEY01 ASC
           END-EXEC.

But for these two

           EXEC SQL AT :DBS
              OPEN CSKEY01N
           END-EXEC.
      *
           EXEC SQL AT :DBS
               FETCH CSKEY01N
               INTO
                 :FLD1,
                 :FLD2
           END-EXEC.

The [Oracle docs](https://docs.oracle.com/cd/A58617_01/server.804/a58232/ch04.htm) state:

> Cursor Control. Cursor control statements such as OPEN, FETCH, and CLOSE are exceptions-they never use an AT clause. If you want to associate a cursor with an explicitly identified database, use the AT clause in the DECLARE CURSOR statement, as follows:
>     EXEC SQL AT :DB-NAME DECLARE emp_cursor CURSOR FOR ... 
>     EXEC SQL OPEN EMP-CURSOR ... 
>     EXEC SQL FETCH EMP-CURSOR ... 
>     EXEC SQL CLOSE EMP-CURSOR END-EXEC.

It is possible I am missing something, of course.

**Comment by GitMensch** [comment]
Ah, that's actually cool and makes sense.
Rechecked with OCESQL: that allows to use `at :DB-NAME` - and then just puts thats to the waste.

For compatibility I suggest to add something like

```yacc
bad_at:
AT dbid { warn("AT DB-NAME is not allowed for CURSOR access, always used from CURSOR DECLARE"); }
| %empty { /* everything fine */ }
;
```

**Comment by GitMensch** [comment]
Rechecked: both ocesql and procob ignore that (without any warning, which I consider bad).

**Comment by mridoni** [comment]
>> Rechecked: both ocesql and procob ignore that (without any warning, which I consider bad), so that's the history of likely many environments having this error in their code.

Implemented as a warning, will be in v1.0.10

**Comment by GitMensch** [comment]
Cool, very useful for legacy code!

**Comment by GitMensch** [comment]
Same in my tests :-)

Of course people are already requesting a `-Wno-cursor-use-with-connection`, `-Werror=cursor-use-with-connection` (or however, but I think those would be nice) - but that's something to track in a different issue - if you want to track that.

**Comment by mridoni** [comment]
Found some problems when used with multiple cursors, reopening the issue, moved the milestone and fixing it. This also needs a better test case, since the current one is not exposing the problem.

**Comment by GitMensch** [comment]
All issues I've tracked so far were either related to #76 or #82, you possibly want to check this before re-visiting multi cursors.

**Comment by mridoni** [comment]
Solved (in my internal repository) but for now only on PostgreSQL.

**Comment by GitMensch** [comment]
That's fine with me - can you create a PR enabling others (ok, me ...) to take this into a local patched version? [or directly post the patch]

**Comment by mridoni** [comment]
Here it is.
[issue44_multiple_conns_with_crsrs.txt](https://github.com/mridoni/gix/files/8341638/issue44_multiple_conns_with_crsrs.txt)
